File: 37. Sudoku Solver.py
class Solution(object):
    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        row = [set() for i in range(9)]
        col = [set() for i in range(9)]
        grid = [[set() for j in range(3)] for i in range(3)]
        si = [str(i) for i in range(10)]
        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    continue
                x = int(board[i][j])
                row[i].add(x)
                col[j].add(x)
                grid[i/3][j/3].add(x)
        
        def solve(k):
            if k == 81:
                return True
            x, y = k/9, k%9
            if board[x][y] != '.':
                return solve(k+1)
            for i in range(1, 10):
                if (i in row[x]) or (i in col[y]) or (i in grid[x/3][y/3]):
                    continue
                row[x].add(i)
                col[y].add(i)
                grid[x/3][y/3].add(i)
                board[x][y] = si[i]
                if solve(k+1):
                    return True
                row[x].remove(i)
                col[y].remove(i)
                grid[x/3][y/3].remove(i)
                board[x][y] = '.'
            return False
        
        solve(0)
        

File: 37. Sudoku Solver.go
func solveSudoku(board [][]byte) {
	rows := [10][10]bool{}
	cols := [10][10]bool{}
	grids := [3][3][10]bool{}

	for i := 0; i < 9; i++ {
		for j := 0; j < 9; j++ {
			k := board[i][j] - '0'
			if k > 0 && k <= 9 {
				rows[i][k] = true
				cols[j][k] = true
				grids[i/3][j/3][k] = true
			}
		}
	}

	var dfs func(int, int) bool
	dfs = func(i, j int) bool {
		if i == 9 {
			return true
		}
		if board[i][j] != '.' {
			return dfs(i+(j+1)/9, (j+1)%9)
		}

		for k := 1; k <= 9; k++ {
			if rows[i][k] || cols[j][k] || grids[i/3][j/3][k] {
				continue
			}
			board[i][j] = byte(k + '0')
			rows[i][k] = true
			cols[j][k] = true
			grids[i/3][j/3][k] = true
			if dfs(i+(j+1)/9, (j+1)%9) {
				return true
			}
			board[i][j] = '.'
			rows[i][k] = false
			cols[j][k] = false
			grids[i/3][j/3][k] = false
		}
		return false
	}
	dfs(0, 0)
}